scanner/indexer/indexer.go

```diff
@@ -46,6 +50,22 @@ func NewIndexer(ctx context.Context) (*Indexer, error) {
 	}
 
 	return &Indexer{
+		fetcher: fetcher,
 		indexer: indexer,
 	}, nil
 }
+
+// Index indexes the given image and returns the index report.
+func (i *Indexer) Index(ctx context.Context, image string, opts ...Option) (*claircore.IndexReport, error) {
+	m, err := i.fetcher.Get(ctx, image, opts...)
```

Comment:
Here we are breaking the contract of the fetcharena; I was expecting it to be primarily a `Realizer` factory, which ClairCore would use to get layers if the index for that layer was not found in the database. Now we give the fetcharena the responsibility of pulling all layers and managing the lifetime of layers downloaded before we check the DB.

It's a bit "hacky". What other alternatives have you considered to reduce the registry round-trip and not break the contracts?

Also, is it worth considering whether it is worth minimizing image layer storage at this point in the Scanner V4? Optimizing registry round-trips is important, but optimizing storage for layers for a single Indexer instance is not as critical considering that once it is indexed, we won't pull it again. Also, on a multi-node indexer, these images will be pulled anyway, while with the database check preventing the pulls, we would not.

Reply:
> Now we give the fetcharena the responsibility of pulling all layers and managing the lifetime of layers downloaded before we check the DB.

This is essentially what ClairCore does, too: https://github.com/quay/claircore/blob/main/libindex/fetcher.go#L39

The only differences are:

* how the layers are fetched
* our fetcher adds the ability to fetches the manifest for the image, which helps with the coordination (the [godoc](https://github.com/quay/claircore/blob/main/libindex/fetcher.go#L39) mentioned FetchArena is meant for coordination)

> Have you considered the possibility that reducing image layer storage in the Scanner V4 may not be as crucial as we initially thought? Optimizing registry round-trips is important, but maybe optimizing storage for layers for a single Indexer instance is not as critical -- specially considering that once it is indexed, we won't pull it again. Also, on a multi-node indexer, these images will be pulled anyway, while with the database check preventing the pulls, we would not, am I right?

ClairCore does, essentially, the same thing as this tries to do. Yes, once the layer is completely indexed and stored in the DB, we will no longer pull the layer. However, there is still the time period before the indexing is complete when ClairCore may pull a layer multiple times. One of the changes our implementation does is preventing this scenario from happening. ClairCore attempts to resolve this, too, but it uses `singleflight`, which only helps deduplicate parallel download calls. It does not prevent the possibility that a layer is downloaded for image A and then later downloaded again for image B before the related layer for image A is added to the DB.

As you mentioned, this does not prevent the case where another instance attempts to pull the same layer during the problematic time period, but it at least ensures no single instance downloads more data than necessary